### PR TITLE
Adding typechain

### PR DESCRIPTION
**Pull request by nicosampler**
Hello everyone, this is Nico speaking. As part of the [BootNode](https://bootnode.dev/) team, I'm working on the development of the direct grants feature.

This specific PR brings [TypeChain](https://github.com/dethcrypto/TypeChain) into the round-manager packages. It's worth mentioning that the process used here can be similarly applied to introduce TypeChain into the other packages.

TypeChain's primary function is to take in abi.json files and generate corresponding Typescript types from them. Beyond this, it also creates factories for these smart contracts, thus enabling type-safe interactions with them.

Given the current usage pattern of abis in grants-stack project, a slight refactor will be required so existing contracts can be consumed by TypeChain. The core of the issue is that the current abis are presently exported as .ts files, not .json files. But this transition can be done progressively by migrating one contract at a time. 

To help you better understand the benefits of TypeChain, I've included a snippet demonstrating the usage of contracts through Typechain.

```tsx
  // import factory
  import { RegistryFactory__factory } from "../../types/generated/typechain";

  // connect contract, at this point the contract already knows the abi and is fully typed. 
  const _registryFactoryContract = registryFactoryContract(chainId);
  const registryFactory = RegistryFactory__factory.connect(
    _registryFactoryContract,
    signerOrProvider
  );
  
  // call contract method.
  const tx = await registryFactory.createRound(
      roundFactoryContract(chainId, true).address as string,
      "0x0d65c0d19f61546198cfebeffc1143ec62ccddbad2a830ebaae2c4c232642b8a", // programId,
      strategyImplementationContract(chainId).address as string,
      encodedRoundParameters,
      encodedStrategyParameters
  );
```

Here is a screenshot showing all the parameters `createRound` receives. 
image

Note: types and factories are generated through the following `scripts` in the `package.json`. 
```
"typechain": "typechain --target=ethers-v5 ./src/features/api/abi/**/*.json --out-dir ./src/types/generated/typechain/",
"postinstall": "pnpm typechain"
```

### CODE REVIEW

**Comment by vacekj** [comment]
Hey Nico, thank you for the PR and excited to be working with you!

We already have a [branch that upgrades wagmi to v1](https://github.com/gitcoinco/grants-stack/pull/1848) and that includes viem. [Viem](http://viem.sh/) leverages [ABIType](https://abitype.dev/) to provide type-safe access to contracts at compile-time without the need to generate anything, by just passing in the ABI as a const object. You can see a brief intro here: https://viem.sh/docs/introduction.html

We will almost certainly be moving forward with integrating viem in the near future, and I'm not sure it would be beneficial to have both in the codebase, as it would lead to a lot of duplication. 

I'm sure we can figure out how to best use typechain or viem to our advantage, so I'd love to have a meeting with both teams and work out the best strategy going forward.

Cheers!

**Comment by nicosampler** [comment]
hi @vacekj, thx for the feedback and the welcome.

Great, didn't know you have already been doing some work to bring type safety to the contracts. I completely agree it does not make sense to add TypeChain and it will be better to wait until #1848 is merged. 

Closing this PR.

**Comment by vacekj** [comment]
@nicosampler sorry for changing our minds so often. after much deliberation, we've decided that due to several blockers, we won't be proceeding with viem at the moment. so you're free to integrate typechain, but please only include your first commit in this PR, and put the rest of the code changes into a new PR. Thanks!

**Comment by nicosampler** [comment]
Hi @vacekj thx for the feedback and the explanation. Regarding the last commit it was unintentional, thx for bringing it up. 

I believe this PR is ready to be reviewed, let me know if there is something else I need to modify.

**Review by vacekj — APPROVED** [review]

File: packages/round-manager/package.json
Comment (nicosampler) [reply packages/round-manager/package.json:134]: oh, it was not intentional, let me check what could have happened. I'll remove it.

Comment (nicosampler) [reply packages/round-manager/package.json]: please check now @vacekj

Comment (vacekj) [reply packages/round-manager/package.json]: Solved, thanks.

**Review by vacekj — APPROVED** [review]

**Review by bard — APPROVED** [review]

**Review by bard — APPROVED** [review]

**Comment by vacekj** [comment]
@nicosampler feel free to merge when you're ready

**Comment by nicosampler** [comment]
@vacekj seems I don't have permission to do so. Can you do it?

**Pull request merged**
